File: leetcode/209.cpp
#include<vector>
using namespace std;

class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        
        //array bigger or equal to target
        //right pointer in the second of array
        //but it could be no result or result 1 or size of array.
        auto right=nums.begin(),left=nums.begin();
        int result=nums.size()+1,sum=nums[0],size=1;
        for(;;){
            if(sum>=target && size<result){
                result=size;
            }
            if(sum>target){
                sum-=*(left++);
                --size;
            }else{
                if(++right!=nums.end()){
                    sum+=*right;
                    ++size;
                }else{
                    break;
                }
                
            }
        }
        if(result==nums.size()+1){
            return 0;
        }
        return result;
    }
};

File: leetcode/209.py
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        p1=-1
        p2=-1
        result=0
        all=0
        while p1<len(nums) and p2<len(nums):
            #print(p1,p2,len(nums))
            if all>=target:
                #print(nums[p1:p2+1])
                if result==0:
                    result=p2-p1
                else:
                    result=min(result,p2-p1)
                p1+=1
                all-=nums[p1]
            else:
                p2+=1
                if p2>=len(nums):
                    break
                all+=nums[p2]

        return result